File: 2024 August/Job Sequencing Problem/main.cpp
// Link: https://www.geeksforgeeks.org/problems/job-sequencing-problem-1587115620/1

#include <iostream>
#include <vector>
#include <algorithm>
#include <set>
using namespace std;

struct Job
{
    int id;     // Job Id
    int dead;   // Deadline of job
    int profit; // Profit if job is over before or on deadline
};

class Solution
{
public:
    vector<int> JobScheduling(Job arr[], int n)
    {
        vector<pair<int, int>> p;
        int maxi = 0;
        for (int i = 0; i < n; i++)
        {
            p.push_back({arr[i].profit, arr[i].dead});
            maxi = max(maxi, arr[i].dead);
        }

        sort(p.rbegin(), p.rend());
        vector<int> vis(maxi + 1, 0);

        set<int> s;
        for (int i = 1; i <= maxi; i++)
        {
            s.insert(i);
        }

        int sum = 0;
        int cnt = 0;

        for (int i = 0; i < p.size(); i++)
        {
            if (s.empty())
                break;
            auto it = s.upper_bound(p[i].second);

            if ((*it) == p[i].second)
            {
                sum += p[i].first;
                cnt++;
                if (it == s.end())
                    it--;
                s.erase(it);
            }
            else if (it != s.begin())
            {
                it--;
                sum += p[i].first;
                cnt++;
                s.erase(it);
            }
        }

        return {cnt, sum};
    }
};

int main()
{
    Solution obj;
    Job arr1[] = {
        {1, 4, 20},
        {2, 1, 10},
        {3, 1, 40},
        {4, 1, 30}};
    int n = sizeof(arr1) / sizeof(Job);
    vector<int> ans = obj.JobScheduling(arr1, n);
    cout << ans[0] << " " << ans[1] << endl;

    Job arr2[] = {
        {1, 2, 100},
        {2, 1, 19},
        {3, 2, 27},
        {4, 1, 25},
        {5, 1, 15}};

    n = sizeof(arr2) / sizeof(Job);
    ans = obj.JobScheduling(arr2, n);
    cout << ans[0] << " " << ans[1] << endl;
    return 0;
};

File: 2024 August/Job Sequencing Problem/main.py
# Link: https://www.geeksforgeeks.org/problems/job-sequencing-problem-1587115620/1


class Job:
    def __init__(self, profit=0, deadline=0):
        self.profit = profit
        self.deadline = deadline
        self.id = 0


class Solution:
    def JobScheduling(self, Jobs, n):
        ans = [0] * n
        abc = []
        for i in Jobs:
            abc.append([i.deadline, i.profit])
        abc.sort(key=lambda x: x[1], reverse=True)
        answer = 0
        for i in abc:
            b, c = i[0], i[1]
            while ans[b - 1] != 0 and b != 0:
                b -= 1
            if b != 0:
                answer += 1
                ans[b - 1] = c
        return [answer, sum(ans)]


if __name__ == "__main__":
    obj = Solution()
    arr1 = [Job(4, 20), Job(1, 10), Job(1, 40), Job(1, 30)]
    n = len(arr1)
    ans = obj.JobScheduling(arr1, n)
    print(ans[0], ans[1])

    arr2 = [Job(2, 100), Job(1, 19), Job(2, 27), Job(1, 25), Job(1, 15)]

    n = len(arr2)
    ans = obj.JobScheduling(arr2, n)
    print(ans[0], ans[1])
